torch/csrc/jit/passes/frozen_linear_folding.cpp

```diff
@@ -0,0 +1,119 @@
+#include <torch/csrc/jit/ir/constants.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/passes/dead_code_elimination.h>
+#include <torch/csrc/jit/passes/fold_linear_bn.h>
+#include <torch/csrc/jit/passes/frozen_linear_folding.h>
+#include <torch/csrc/jit/passes/utils/optimization_utils.h>
+
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/Functions.h>
+#else
+#include <ATen/ops/ones_like.h>
+#include <ATen/ops/zeros_like.h>
+#endif
+
+namespace torch {
+namespace jit {
+
+namespace {
+
+using Tensor = at::Tensor;
+
+bool supportedLinearNode(Node* n) {
+  if (n->kind() == aten::linear) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+bool FoldFrozenLinearBatchnorm(Block* b) {
+  bool graph_modified = false;
+  for (Node* n : b->nodes()) {
+    for (Block* block : n->blocks()) {
+      graph_modified |= FoldFrozenLinearBatchnorm(block);
+    }
+
+    if (n->kind() == aten::batch_norm &&
+        supportedLinearNode(n->inputs().at(0)->node())) {
+      auto linear = n->inputs().at(0)->node();
+      auto bn = n;
+
+      if (nonConstantParameters(linear) || nonConstantParameters(bn)) {
+        continue;
+      }
+
+      auto bn_rm_ivalue = bn->namedInput("running_mean");
+      auto bn_rv_ivalue = bn->namedInput("running_var");
+
+      // check running_mean and running_var has value, if they are
+      // None(track_running_stats=False), skiping the folding path.
+      if (bn_rm_ivalue->type() == NoneType::get() &&
+          bn_rv_ivalue->type() == NoneType::get()) {
+        continue;
+      }
+
+      auto bn_rm = constant_as<Tensor>(bn->namedInput("running_mean")).value();
+      auto bn_rv = constant_as<Tensor>(bn->namedInput("running_var")).value();
+      auto bn_eps = constant_as<double>(bn->namedInput("eps")).value();
+      auto linear_w = constant_as<Tensor>(linear->namedInput("weight")).value();
+
+      // implementation taken from torch/nn/utils/fusion.py
+      Tensor linear_b;
+      if (linear->namedInput("bias")->type() == NoneType::get()) {
```

Comment:
this looks like it's mostly copied from frozen_conv_folding.cpp, is that accurate? recently we had some issues with autocasting, see https://github.com/pytorch/pytorch/pull/77617. Can you add some tests like this to confirm that it's not an issue in this case?

Reply:
> this looks like it's mostly copied from frozen_conv_folding.cpp, is that accurate?

That's correct, this is mostly copied from frozen_conv_folding.cpp. 

> recently we had some issues with autocasting, see #77617

Thanks for the notice. I verified that similar issue occurs for jit autocasting for lin-bn1d only and on gpu only (inputs are only casted to half on gpu). 

The issue occurs with {2,3}d input tensor, which is the correct usage of bn1d; however, the issue will also occur for an incorrect usage of bn{2,3}d with {2,3}d input tensor. Below, replace `nn.BatchNorm1d(out_features)` with `nn.BatchNorm{2,3}d(out_features)`.

To reproduce `RuntimeError: self and mat2 must have the same dtype`:
```
import torch
import torch.nn as nn

class LinearBN(torch.nn.Module):
    def __init__(self, in_features, out_features):
        super(LinearBN, self).__init__()
        self.linear = nn.Linear(in_features, out_features, bias=False)
        self.bn = nn.BatchNorm1d(out_features)
    
    def forward(self, x):
        x = self.linear(x)
        return self.bn(x)

net = LinearBN(32, 32).cuda().eval()  
print(net.linear.weight.dtype) #torch.float32
img = torch.rand((1, 32, 32)).cuda()
with torch.cuda.amp.autocast(True):
    net = torch.jit.script(net)
    net = torch.jit.freeze(net)
print(net.graph)
lin_node = net.graph.findNode("aten::linear", True)
print(lin_node.namedInput("weight").type().dtype()) #torch.float16
print(lin_node.namedInput("bias").type().dtype()) #torch.float32
net(img)
```